benchbuild/experiment.py

```diff
@@ -124,49 +130,56 @@
         return type(self).SCHEMA
 
     @schema.validator
-    def validate_schema(self, _, new_schema):
+    def validate_schema(self, _: tp.Any, new_schema) -> bool:
         if new_schema is None:
             return True
         if isinstance(new_schema, collections.abc.Iterable):
             return True
         return False
 
     @abstractmethod
-    def actions_for_project(self, project):
+    def actions_for_project(self, project: Project) -> Actions:
         """
         Get the actions a project wants to run.
 
         Args:
             project (benchbuild.Project): the project we want to run.
         """
 
-    def actions(self):
+    def actions(self) -> Actions:
         """
         Common setup required to run this experiment on all projects.
         """
         actions = []
 
-        for project in self.projects:
-            prj_cls = self.projects[project]
-
-            prj_actions = []
-            for version in self.sample(prj_cls, prj_cls.versions()):
-                p = prj_cls(self, version=version)
-
-                atomic_actions = [
-                    Clean(p),
-                    MakeBuildDir(p),
-                    Echo(message="Selected {0} with version {1}".format(
-                        p.name, p.version)),
-                    Containerize(obj=p, actions=self.actions_for_project(p))
+        for prj_cls in self.projects:
+            prj_actions: Actions = []
+
+            project_variants = source.product(*prj_cls.SOURCE)
+            for variant in project_variants:
+                var_context = source.context(*variant)
+                version_str = source.to_str(*variant)
+
+                p = prj_cls(self, variant=var_context)
+                atomic_actions: Actions = [
+                    actns.Clean(p),
+                    actns.MakeBuildDir(p),
+                    actns.Echo(message="Selected {0} with version {1}".format(
+                        p.name, version_str)),
+                    actns.ProjectEnvironment(p),
+                    actns.Containerize(obj=p,
+                                       actions=self.actions_for_project(p))
                 ]
-                prj_actions.append(RequireAll(actions=atomic_actions))
+
+                prj_actions.append(actns.RequireAll(actions=atomic_actions))
             actions.extend(prj_actions)
 
-        actions.append(CleanExtra(self))
+        actions.append(actns.CleanExtra(self))
         return actions
 
-    def sample(self, prj_cls, versions=None):
+    def sample(self,
```

Comment:
Codacy found an issue: [Method could be a function](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5694346)